File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while r >= l:
            i = (r + l) // 2
            if nums[i] == target:
                return i
            elif target > nums[i]:
                l = i + 1
            else:
                r = i - 1
        return -1

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {

    private void discover(int[] nums, List<Integer> comb, int comb_sum, int cursor, List<List<Integer>> res, int target) {
        if (comb_sum == target) { 
            res.add(new ArrayList(comb));
            return;
        }

        for(int i = cursor; i < nums.length; i++) {
            if (comb_sum + nums[i] <= target) {
                comb_sum += nums[i];
                comb.add(nums[i]);
                discover(nums, comb, comb_sum, i, res, target);
                comb_sum -= nums[i];
                comb.removeLast();
            }
        }
    }

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        
        List<List<Integer>> res = new ArrayList();

        discover(nums, new ArrayList(), 0, 0, res, target);

        return res;
    }
}
